pkg/cmd/version/version.go

```diff
@@ -18,21 +18,133 @@ package version
+				cmd.CheckError(err)
+
+				arkClient = client.ArkV1()
+			}
+
+			printVersion(os.Stdout, clientOnly, arkClient, serverStatusGetter)
 		},
 	}
 
+	c.Flags().DurationVar(&serverStatusGetter.timeout, "timeout", serverStatusGetter.timeout, "maximum time to wait for server version to be reported")
+	c.Flags().BoolVar(&clientOnly, "client-only", clientOnly, "only get ark client version, not server version")
+
 	return c
 }
+
+func printVersion(w io.Writer, clientOnly bool, client arkv1client.ServerStatusRequestsGetter, serverStatusGetter serverStatusGetter) {
+	fmt.Fprintln(w, "Client:")
+	fmt.Fprintf(w, "\tVersion: %s\n", buildinfo.Version)
+	fmt.Fprintf(w, "\tGit commit: %s\n", buildinfo.FormattedGitSHA())
+
+	if clientOnly {
+		return
+	}
+
+	serverStatus, err := serverStatusGetter.getServerStatus(client)
+	if err != nil {
+		fmt.Fprintf(w, "<error getting server version: %s>\n", err)
+		return
+	}
+
+	fmt.Fprintln(w, "Server:")
+	fmt.Fprintf(w, "\tVersion: %s\n", serverStatus.Status.ServerVersion)
+}
+
+type serverStatusGetter interface {
+	getServerStatus(client arkv1client.ServerStatusRequestsGetter) (*arkv1api.ServerStatusRequest, error)
+}
+
+type defaultServerStatusGetter struct {
+	namespace string
+	timeout   time.Duration
+}
+
+func (g *defaultServerStatusGetter) getServerStatus(client arkv1client.ServerStatusRequestsGetter) (*arkv1api.ServerStatusRequest, error) {
+	req := serverstatusrequest.NewBuilder().Namespace(g.namespace).GenerateName("ark-cli-").Build()
+
+	created, err := client.ServerStatusRequests(g.namespace).Create(req)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	defer client.ServerStatusRequests(g.namespace).Delete(created.Name, nil)
+
+	listOptions := metav1.ListOptions{
+		// TODO: once the minimum supported Kubernetes version is v1.9.0, uncomment the following line.
+		// See http://issue.k8s.io/51046 for details.
+		//FieldSelector:   "metadata.name=" + req.Name
+		ResourceVersion: created.ResourceVersion,
+	}
+	watcher, err := client.ServerStatusRequests(g.namespace).Watch(listOptions)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	defer watcher.Stop()
+
+	expired := time.NewTimer(g.timeout)
+	defer expired.Stop()
+
+Loop:
+	for {
+		select {
+		case <-expired.C:
+			return nil, errors.New("timed out waiting for server status request to be processed")
+		case e := <-watcher.ResultChan():
+			updated, ok := e.Object.(*arkv1api.ServerStatusRequest)
+			if !ok {
+				return nil, errors.Errorf("unexpected type %T", e.Object)
+			}
+
+			// TODO: once the minimum supported Kubernetes version is v1.9.0, remove the following check.
+			// See http://issue.k8s.io/51046 for details.
+			if updated.Name != created.Name {
+				continue
+			}
+
+			switch e.Type {
+			case watch.Deleted:
+				return nil, errors.New("download request was unexpectedly deleted")
+			case watch.Modified:
+				if updated.Status.Phase == arkv1api.ServerStatusRequestPhaseProcessed && updated.Status.ServerVersion != "" {
```

Comment:
Don't we want to break out of the loop if the phase is processed, regardless of the value of `updated.Status.ServerVersion`?